### PR TITLE
plutus-ledger-api V2

### PR DESCRIPTION
**Pull request by szg251**
- [x] Move ToCSL/FromCSL to pla
- [x] Fix Value serde serialization (#34)
- [x] Add Display instances where it makes sense (#51)
- [x] Display instance golden tests
- [x] `from_bytes` implementations with invariant checking 
- [x] V3 types
- [x] Derivation macro for `IsPlutusData`

### CODE REVIEW

File: plutus-ledger-api/src/v1/value.rs [hunk 1]
```diff
@@ -78,12 +145,94 @@ impl Json for CurrencySymbol {
     }
 }
 
+pub(crate) fn currency_symbol(input: &str) -> IResult<&str, CurrencySymbol, VerboseError<&str>> {
+    context(
+        "currency symbol",
+        alt((
```
Comment (chfanghr) [inline plutus-ledger-api/src/v1/value.rs:151]: Why not:

```rust
map_res(ledger_bytes, |(LedgerBytes(bytes))| Self::from_bytes(bytes))
```
 
`alt` is kinda expensive in the sense that in the worst case, we have to execute both parsers needlessly

File: plutus-ledger-api/src/v1/value.rs [hunk 2]
```diff
@@ -30,22 +55,64 @@ pub enum CurrencySymbol {
...
+            .map(|(_, cs)| cs)
+    }
+}
+
 impl IsPlutusData for CurrencySymbol {
     fn to_plutus_data(&self) -> PlutusData {
         match self {
-            CurrencySymbol::Ada => String::from("").to_plutus_data(),
             CurrencySymbol::NativeToken(policy_hash) => policy_hash.to_plutus_data(),
+            CurrencySymbol::Ada => String::from("").to_plutus_data(),
```
Comment (chfanghr) [inline plutus-ledger-api/src/v1/value.rs:106]: I kinda hate this. We are relying on the internal implementation of `IsPlutusData` instance of `String` while `CurrencySymbol` is in fact not a `String`.

File: plutus-ledger-api/src/v1/value.rs [hunk 3]
```diff
@@ -186,11 +355,79 @@ impl Value {
...
+    let (input, cs) = currency_symbol(input)?;
+
+    let (input, tn) = alt((preceded(char('.'), token_name), success(TokenName::ada())))(input)?;
+
+    let (input, amount) = preceded(char(' '), big_int)(input)?;
+
+    Ok((input, (cs, tn, amount)))
+}
+
+pub(crate) fn value(input: &str) -> IResult<&str, Value, VerboseError<&str>> {
```
Comment (chfanghr) [inline plutus-ledger-api/src/v1/value.rs:415]: Leave some doc strings for the parsers please: the formats of strings they expect? some examples? It doesn't need to be formal but should help people understand what's happening at a glance.

File: plutus-ledger-api/src/v3/mod.rs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+// Inherited from v2
+pub use crate::v2::address;
+pub use crate::v2::assoc_map;
+pub use crate::v2::crypto;
+pub use crate::v2::datum;
+pub use crate::v2::interval;
+pub use crate::v2::redeemer;
+pub use crate::v2::script;
+pub use crate::v2::value;
```
Comment (chfanghr) [inline plutus-ledger-api/src/v3/mod.rs:15]: Oh we have to reexport `TransactionInput` from v1 and `TransactionOutput` from v2 in the `transaction` module

File: plutus-ledger-api/tests/display.proptest-regressions
Comment (chfanghr) [inline plutus-ledger-api/tests/display.proptest-regressions:1]: Why is this committed? Did we resolve this?

**Review by chfanghr — CHANGES_REQUESTED** [review]
HUGE

File: plutus-ledger-api/CHANGELOG.md [hunk 5]
```diff
@@ -4,13 +4,23 @@
 This changelog is based on [Keep A
 Changelog](https://keepachangelog.com/en/1.1.0).
 
-## Unreleased
+## v2.0.0
 
 ### Added
 
+- Added cardano-serialization-lib conversion traits (`ToCSL` and `FromCSL`)
```
Comment (chfanghr) [inline plutus-ledger-api/CHANGELOG.md:11]: Link to this pr

File: plutus-ledger-api/src/v1/value.rs [hunk 6]
```diff
@@ -143,6 +292,26 @@ impl Value {
...
+            // Has negative entries?
+            .filter(|_, _, amount| amount < &BigInt::from(0u32))
+            .is_empty()
+    }
+
+    pub fn is_pure_ada(self) -> bool {
+        let inner = self.normalize().0;
+        let inner: Vec<_> = inner.into_iter().collect();
+
+        match inner.as_slice() {
```
[the remark this answers is not in the document]
Comment (szg251) [reply plutus-ledger-api/src/v1/value.rs:308]: I refactored this whole thing into something simpler, without the need of cloning and normalizing the whole Value

File: plutus-ledger-api/src/v1/value.rs (hunk 1 above)
Comment (szg251) [reply plutus-ledger-api/src/v1/value.rs:151]: good idea, let's roll with this

File: plutus-ledger-api/src/v1/value.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (chfanghr) [reply plutus-ledger-api/src/v1/value.rs:308]: I think you replied to the wrong comment lol

File: plutus-ledger-api/src/v1/value.rs (hunk 2 above)
Comment (szg251) [reply plutus-ledger-api/src/v1/value.rs:106]: I just gave it a `PlutusData::Bytes(Vec::with_capacity(0))` directly

File: plutus-ledger-api/src/aux.rs [hunk 7]
```diff
@@ -75,6 +80,8 @@ pub fn guard_bytes(ctx: &str, bytes: Vec<u8>, expected: usize) -> Result<Vec<u8>
     }
 }
 
+/// Nom parser for BigInt
+/// Expects an arbitrary length integer, optionally signed
```
Comment (chfanghr) [inline plutus-ledger-api/src/aux.rs:84]: decimal numbers

File: plutus-ledger-api/tests/display.proptest-regressions
Comment (chfanghr) [inline plutus-ledger-api/tests/display.proptest-regressions:1]: Oh no it's coming back

**Review by chfanghr — APPROVED** [review]
Amazing doc strings! Thanks!

**Pull request merged**